text/0000-additional-exported-layers.md

```diff
@@ -0,0 +1,123 @@
+
+# Definitions
+[definitions]: #definitions
+
+- `Workspace`: The default application directory where the application source code is mounted and final build outputs are written. This directory is exported.
+- `Volume`: A set of directories describing where the buildpack process is likely to write data specific and should be exported alongside the application workspace. 
+- `WorkingDir`: The default working directory for a container.
+
+# Motivation
+[motivation]: #motivation
+
+The main motivation behind this RFC is to unlock exporting application images that require libraries or software to be installed in certain directories apart from the default application directory and these installations are rebase safe and self-contained. This unlocks use cases where buildpacks may install software in `/opt` directories (for AWS Lambda extensions or other common standalone software) for example or preserve some settings in a `/home/$USER/.config` for the output application. This provides a rebase safe extension point without major changes to the buildpacks API.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Users would be able to define stack images with `VOLUME` directives in Dockerfiles.
+
+
+For example a build/run base image may look like - 
+
+```Dockerfile
+# Set a common base
+FROM ubuntu:bionic as base
+
+# Set required CNB information
+ENV CNB_USER_ID=1000
+ENV CNB_GROUP_ID=1000
+ENV CNB_STACK_ID="io.buildpacks.samples.stacks.bionic"
+LABEL io.buildpacks.stack.id="io.buildpacks.samples.stacks.bionic"
+
+# Create the user
+RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
+  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb
+
+# Start a new run stage
+FROM base as run
+
+# Volume instructions to identify paths that the lifecycle
+# may safely overlay on the run image.
+# This would ensure that these paths are rebase safe as
+# docker currently discards any changes made to directories
+# declared as volumes. See https://docs.docker.com/engine/reference/builder/#notes-about-specifying-volumes
+VOLUME ["/app", "/opt/extensions", "/home/cnb/.config"]
+
+# Set user and group (as declared in base image)
+USER ${CNB_USER_ID}:${CNB_GROUP_ID}
+
+# Start a new build stage
+FROM base as build
+
+# Volume instruction in the build image that the lifecycle
+# will export out. This must match the volumes declared in the run image.
+VOLUME ["/app", "/opt/extensions", "/home/cnb/.config"]
+
+# Optionally users can override the default application workspace
+# by specifying it through the WORKDIR directive in the build
+# and run images.
+WORKDIR /app
+
+# Set user and group (as declared in base image)
+USER ${CNB_USER_ID}:${CNB_GROUP_ID}
+```
+
+Buildpacks will be passed a list of volumes through the `CNB_EXPORT_VOLUMES` variable and this variable would be a JSON list. Buildpacks may read this environment variable during `detect` or `build`.
```

Comment:
```suggestion
Buildpacks will be passed a list of volumes through the `CNB_EXPORT_DIRS` variable and this variable would be a JSON list. Buildpacks may read this environment variable during `detect` or `build`.
```
Given we are exporting the contents of these build-time volumes into real layers in the resulting app image maybe this makes more sense from the buildpack perspective? It is irrelevent to the buildpack that these are build time volumes.